scbuild: implement RENAME CONSTRAINT in declarative schema changer

rafiss opened the pull request:
This patch implements full support for RENAME CONSTRAINT in the
declarative schema changer. The implementation handles two types of
constraints:

1. Index-backed constraints (PRIMARY KEY, UNIQUE with index): These
   rename by dropping the old IndexName element and adding a new one
   with the updated name.

2. Non-index constraints (CHECK, FK, UNIQUE without index): These
   rename by dropping the old ConstraintWithoutIndexName element and
   adding a new one with the updated name.

The implementation includes proper validation:
- Checks for constraint existence
- Validates the constraint is not being added/dropped concurrently
- Prevents name conflicts with existing constraints and indexes
- Ensures views don't depend on index-backed constraints being renamed

New dependency rules ensure the old name element reaches ABSENT before
the new name element becomes PUBLIC, preventing conflicts and ensuring
correct rollback behavior.

###  scplan: add dependency rules for cross-constraint name conflicts during rename

This commit enhances the RENAME CONSTRAINT implementation by adding
dependency rules to handle name conflicts that occur when multiple
constraints are renamed in the same schema change operation.

Previously, the dependency rules only handled simple single-constraint
renames (A → B). However, when multiple constraints are renamed such
that names are swapped or reused (e.g., C1: "old" → "new" while C2:
"new" → "other"), the original rules could allow conflicting names to
exist simultaneously during execution.

This adds two new Precedence rules:

1. For index-backed constraints: Ensures old IndexName elements reach
   ABSENT before new IndexName elements with the same name become PUBLIC.
   Includes special handling to exclude secondary index swaps during
   ALTER PRIMARY KEY operations.

2. For cross-type conflicts: Handles name conflicts between any
   combination of ConstraintWithoutIndexName and IndexName elements,
   preventing CHECK constraints from conflicting with UNIQUE constraint
   names, for example.

---

fixes #148341

Release note: None

rafiss committed:
scbuild: implement RENAME CONSTRAINT in declarative schema changer

This patch implements full support for RENAME CONSTRAINT in the
declarative schema changer. The implementation handles two types of
constraints:

1. Index-backed constraints (PRIMARY KEY, UNIQUE with index): These
   rename by dropping the old IndexName element and adding a new one
   with the updated name.

2. Non-index constraints (CHECK, FK, UNIQUE without index): These
   rename by dropping the old ConstraintWithoutIndexName element and
   adding a new one with the updated name.

The implementation includes proper validation:
- Checks for constraint existence
- Validates the constraint is not being added/dropped concurrently
- Prevents name conflicts with existing constraints and indexes
- Ensures views don't depend on index-backed constraints being renamed

New dependency rules ensure the old name element reaches ABSENT before
the new name element becomes PUBLIC, preventing conflicts and ensuring
correct rollback behavior.

Resolves: #148341
Epic: CRDB-31465

Release note: None

spilchen reviewed (commented):
Looks good. Just had a few minor questions.

> @spilchen reviewed 2 of 17 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-:-Od99iGGCxVC14cCUKHz:bwhkg58)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/schemachanger/scbuild/testdata/alter_table_rename_constraint` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od92jIX97QCAJ6Jcmfp:-Od92jIX97QCAJ6Jcmfq:b-pm1gi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scbuild/testdata/alter_table_rename_constraint#L57)):*
> ```
> 
> build
> ALTER TABLE t5 RENAME CONSTRAINT con1 TO con1_old, RENAME CONSTRAINT con2 TO con1;
> ```

do we need a rule to ensure we execute the first rename before the second? We have rules to ensure con1 is dropped before we add con1_old. And same for con2 drop and con1 add. But do we need a rule to ensure ordering between the two renames?
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 1450 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od97_ND5RqORcmaiA6q:-Od97_ND5RqORcmaiA6r:bcv3i2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scbuild/builder_state.go#L1450)):*
> ```Go
> 		constraintIDMatches := constraintIDAttr != nil && constraintIDAttr.(catid.ConstraintID) == constraintID
> 
> 		// For index-backed constraints, we also want to include the IndexName.
> ```

nit: should we check that the element we match here is an `scpb.IndexName`? I feel like this could end up matching a lot of things, since IndexID is a fairly common attribute.


<!-- Sent from Reviewable.io -->

fqazi reviewed (changes requested):
Some pretty minor comments / concerns.

> @fqazi reviewed 16 of 17 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-:-Od9ewuq11sp77lMjBUJ:bewn5os)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/testdata/alter_table_rename_constraint` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od92jIX97QCAJ6Jcmfp:-Od9ehqw0X-h6HiNTjj0:b-tlzpia) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scbuild/testdata/alter_table_rename_constraint#L57)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

do we need a rule to ensure we execute the first rename before the second? We have rules to ensure con1 is dropped before we add con1_old. And same for con2 drop and con1 add. But do we need a rule to ensure ordering between the two renames?
</blockquote></details>

Think we need a rule to ensure the old named is freed first? They don't have to be in order, just the names are freed correctly.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_rename_constraint.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od9e9YKAPkbiUUCZBqE:-Od9e9YKAPkbiUUCZBqF:b9ly4rd) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scplan/internal/rules/current/dep_rename_constraint.go#L23)):*
> ```Go
> 	registerDepRule(
> 		"old index name is dropped before new index name is added",
> 		scgraph.Precedence,
> ```

Do we want to be stricter here with SameStagePrecendence? Otherwise, there is a risk there exists such a stage where neither the old or new name works. 
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_rename_constraint.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od9eRMR2H9OiVVQ7Z6G:-Od9eRMR2H9OiVVQ7Z6H:b-3q4j7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scplan/internal/rules/current/dep_rename_constraint.go#L41)):*
> ```Go
> 	registerDepRule(
> 		"old constraint name is dropped before new constraint name is added",
> 		scgraph.Precedence,
> ```

Similar comment here.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_rename_constraint.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od9dQh7B_JpzjNJaC58:-Od9dQh7B_JpzjNJaC59:b-3ziuh5) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_rename_constraint.go#L41)):*
> ```Go
> 
> 	// Handle both index-backed and non-index constraints.
> 	_, _, indexName := scpb.FindIndexName(constraintElems)
> ```

Nit: Lets use non-deprecated API here and just do FilterIndexName().MustGetOne()

Similar comment on the next line.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_rename_constraint.go` line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od9dkCE7BfozEk9MQtH:-Od9dkCE7BfozEk9MQtI:bnd21os) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_rename_constraint.go#L89)):*
> ```Go
> 	// Check ConstraintWithoutIndexName elements (CHECK, FK, UNIQUE without index).
> 	tableElems := b.QueryByID(tableID)
> 	scpb.ForEachConstraintWithoutIndexName(tableElems, func(
> ```

Nit: Let's switch to the filter API. Similar to other spots.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
Also great work @rafiss!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-:-Od9f1Qx-Ps-4bv5nZcQ:b-w3e0uh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)



<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154608)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 1450 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od97_ND5RqORcmaiA6q:-Od9jtw06u9onWIsImUr:b-fyk56s) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scbuild/builder_state.go#L1450)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: should we check that the element we match here is an `scpb.IndexName`? I feel like this could end up matching a lot of things, since IndexID is a fairly common attribute.
</blockquote></details>

i don't think that's what we want here -- IMO this should actually return all the elements that pertain to that index, like `scpb.PrimaryIndex` and `scpb.SecondaryIndex`, and the consumer of this value should look for the element they want. i'll adjust this comment to clarify.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_rename_constraint.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od9e9YKAPkbiUUCZBqE:-OdA7ywi9Dye7eLT0gm9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scplan/internal/rules/current/dep_rename_constraint.go#L23)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Do we want to be stricter here with SameStagePrecendence? Otherwise, there is a risk there exists such a stage where neither the old or new name works.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_rename_constraint.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od9eRMR2H9OiVVQ7Z6G:-OdA7zJ6-35XpmhsgDNl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scplan/internal/rules/current/dep_rename_constraint.go#L41)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Similar comment here.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/testdata/alter_table_rename_constraint` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od92jIX97QCAJ6Jcmfp:-OdA80ve7_NeHYaKLVXB:b-lar5pd) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scbuild/testdata/alter_table_rename_constraint#L57)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Think we need a rule to ensure the old named is freed first? They don't have to be in order, just the names are freed correctly.
</blockquote></details>

this rule makes sense, but looks like it breaks the ALTER PK case. i'm brainstorming ways to resolve this.
```
        /Users/rafiss/.cache/bazel/_bazel_rafiss/bases/cf10ff/1/sandbox/darwin-sandbox/2088/execroot/com_github_cockroachdb_cockroach/bazel-out/darwin_arm64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/alter_table:1961:
        expected success, but found
        (XX000) internal error: building declarative schema changer plan in StatementPhase (rollback=false) for ALTER TABLE: validate graph: graph is not acyclical
        graph.go:366: in func1()
        DETAIL: cycle:
        [[SecondaryIndex:{DescID: 170, IndexID: 2, ConstraintID: 1, RecreateSourceIndexID: 0, RecreateTargetIndexID: 0}, ABSENT], WRITE_ONLY] --> op edge
        [[SecondaryIndex:{DescID: 170, IndexID: 2, ConstraintID: 1, RecreateSourceIndexID: 0, RecreateTargetIndexID: 0}, ABSENT], DELETE_ONLY] --> index no longer public before index name
        [[IndexName:{DescID: 170, Name: t54629_c_key, IndexID: 2}, ABSENT], ABSENT] --> old index name is dropped before conflicting new index name is added
        [[IndexName:{DescID: 170, Name: t54629_c_key, IndexID: 3}, PUBLIC], PUBLIC] --> index dependents exist before index becomes public
        [[SecondaryIndex:{DescID: 170, IndexID: 3, ConstraintID: 3, TemporaryIndexID: 4, SourceIndexID: 1, RecreateSourceIndexID: 2, RecreateTargetIndexID: 5}, PUBLIC], PUBLIC] --> replacement secondary index should be validated before the old one becomes invisible
        [[SecondaryIndex:{DescID: 170, IndexID: 2, ConstraintID: 1, RecreateSourceIndexID: 0, RecreateTargetIndexID: 0}, ABSENT], VALIDATED] --> op edge
        [[SecondaryIndex:{DescID: 170, IndexID: 2, ConstraintID: 1, RecreateSourceIndexID: 0, RecreateTargetIndexID: 0}, ABSENT], WRITE_ONLY]
```


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> @spilchen reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154608)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 1450 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od97_ND5RqORcmaiA6q:-OdE21Mg38_7oAFs5n9T:b-jdrnqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scbuild/builder_state.go#L1450)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't think that's what we want here -- IMO this should actually return all the elements that pertain to that index, like `scpb.PrimaryIndex` and `scpb.SecondaryIndex`, and the consumer of this value should look for the element they want. i'll adjust this comment to clarify.
</blockquote></details>

Got it. I think I got thrown by the comment saying we want to include IndexName also. 


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154608)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/testdata/alter_table_rename_constraint` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-Od92jIX97QCAJ6Jcmfp:-OdECfc-E-0wqfD9kOth:b-qblhn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/70aa30d51caa27e2ea173922a2977530ece69287/pkg/sql/schemachanger/scbuild/testdata/alter_table_rename_constraint#L57)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this rule makes sense, but looks like it breaks the ALTER PK case. i'm brainstorming ways to resolve this.
```
        /Users/rafiss/.cache/bazel/_bazel_rafiss/bases/cf10ff/1/sandbox/darwin-sandbox/2088/execroot/com_github_cockroachdb_cockroach/bazel-out/darwin_arm64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/alter_table:1961:
        expected success, but found
        (XX000) internal error: building declarative schema changer plan in StatementPhase (rollback=false) for ALTER TABLE: validate graph: graph is not acyclical
        graph.go:366: in func1()
        DETAIL: cycle:
        [[SecondaryIndex:{DescID: 170, IndexID: 2, ConstraintID: 1, RecreateSourceIndexID: 0, RecreateTargetIndexID: 0}, ABSENT], WRITE_ONLY] --> op edge
        [[SecondaryIndex:{DescID: 170, IndexID: 2, ConstraintID: 1, RecreateSourceIndexID: 0, RecreateTargetIndexID: 0}, ABSENT], DELETE_ONLY] --> index no longer public before index name
        [[IndexName:{DescID: 170, Name: t54629_c_key, IndexID: 2}, ABSENT], ABSENT] --> old index name is dropped before conflicting new index name is added
        [[IndexName:{DescID: 170, Name: t54629_c_key, IndexID: 3}, PUBLIC], PUBLIC] --> index dependents exist before index becomes public
        [[SecondaryIndex:{DescID: 170, IndexID: 3, ConstraintID: 3, TemporaryIndexID: 4, SourceIndexID: 1, RecreateSourceIndexID: 2, RecreateTargetIndexID: 5}, PUBLIC], PUBLIC] --> replacement secondary index should be validated before the old one becomes invisible
        [[SecondaryIndex:{DescID: 170, IndexID: 2, ConstraintID: 1, RecreateSourceIndexID: 0, RecreateTargetIndexID: 0}, ABSENT], VALIDATED] --> op edge
        [[SecondaryIndex:{DescID: 170, IndexID: 2, ConstraintID: 1, RecreateSourceIndexID: 0, RecreateTargetIndexID: 0}, ABSENT], WRITE_ONLY]
```
</blockquote></details>

We could apply IsNotPotentialSecondaryIndexSwap here. The PK swap case has its own way of ordering this correctly.


<!-- Sent from Reviewable.io -->

rafiss committed:
scplan: add dependency rules for cross-constraint name conflicts during rename

This commit enhances the RENAME CONSTRAINT implementation by adding
dependency rules to handle name conflicts that occur when multiple
constraints are renamed in the same schema change operation.

Previously, the dependency rules only handled simple single-constraint
renames (A → B). However, when multiple constraints are renamed such
that names are swapped or reused (e.g., C1: "old" → "new" while C2:
"new" → "other"), the original rules could allow conflicting names to
exist simultaneously during execution.

This adds two new Precedence rules:

1. For index-backed constraints: Ensures old IndexName elements reach
   ABSENT before new IndexName elements with the same name become PUBLIC.
   Includes special handling to exclude secondary index swaps during
   ALTER PRIMARY KEY operations.

2. For cross-type conflicts: Handles name conflicts between any
   combination of ConstraintWithoutIndexName and IndexName elements,
   preventing CHECK constraints from conflicting with UNIQUE constraint
   names, for example.

Release note: None

fqazi reviewed (approved):
Nice @rafiss <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> @fqazi reviewed 5 of 5 files at r2, 40 of 40 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154608#-:-OdUDMNv6xGnjKlyzYIo:b5elt7e)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)



<!-- Sent from Reviewable.io -->
